lib/policies/oauth2/site.js

```diff
@@ -1,13 +1,25 @@
-'use strict';
-
 const passport = require('passport');
 const path = require('path');
+const services = require('../../services/index');
 
 module.exports.loginForm = (request, response) => response.render(path.join(__dirname, 'views/login'));
 
 module.exports.login = passport.authenticate('local', { successReturnToOrRedirect: '/', failureRedirect: '/login' });
 
-module.exports.logout = (request, response) => {
-  request.logout();
-  response.redirect('/');
+module.exports.logout = (req, res) => {
+  req.logout();
+  if (!req.query.returnTo) {
+    return res.redirect('/'); // TODO: implement settings section for default view path and return redirects
+  }
+  const returnUrl = decodeURI(req.query.returnTo);
+  if (!req.query.clientId) {
+    return res.redirect(returnUrl);
+  }
+
+  const client = services.application.get(req.query.clientId);
+  if (client && client.redirectUri && client.redirectUri === returnUrl) {
```

Comment:
This is returning a `Promise` that should be resolved before accessing its data. These lines will probably make EG crash.